.github/workflows/pr_checks.yml

```
@@ -0,0 +1,245 @@
+name: PR Checks
+on:
+  pull_request:
+  push:
+    branches:
+      - master
+    tags: '*'
+concurrency:
+  # Skip intermediate builds: all builds except for builds on the `master` branch
+  # Cancel intermediate builds: only pull request builds
+  group: ${{ github.workflow }}-${{ github.ref }}-${{ github.ref != 'refs/heads/master' || github.run_number }}
+  cancel-in-progress: ${{ startsWith(github.ref, 'refs/pull/') }}
+permissions:
+  contents: read
+jobs:
+  finalize-pr-checks:
+    if: always() # this line is important to keep the `finalize` job from being marked as skipped; do not change or delete this line
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    needs: [
+      checked-in-files,
+      build,
+      npm-run-test,
+      make-targets,
+      stalecheck-npm-install,
+    ]
+    steps:
+      - run: |
+          echo checked-in-files: ${{ needs.checked-in-files.result }}
+          echo build: ${{ needs.build.result }}
+          echo npm-run-test: ${{ needs.npm-run-test.result }}
+          echo make-targets: ${{ needs.make-targets.result }}
+          echo stalecheck-npm-install: ${{ needs.stalecheck-npm-install.result }}
+      # The last line must NOT end with ||
+      # All other lines MUST end with ||
+      - run: exit 1
+        if: |
+          (needs.checked-in-files.result != 'success') ||
+          (needs.build.result != 'success') ||
+          (needs.npm-run-test.result != 'success') ||
+          (needs.make-targets.result != 'success') ||
+          (needs.stalecheck-npm-install.result != 'success')
```

Logically equivalent. Honestly though you can just remove this block entirely as any failed job will still cause this workflow to overall fail.
```suggestion
      - run: exit 1
        if: ${{ !contains(needs.*.result, 'failure') || !contains(needs.*.result, 'cancelled') || !contains(needs.*.result, 'skipped') }}
```

---

> Honestly though you can just remove this block entirely as any failed job will still cause this workflow to overall fail.

The idea here is that we want to use GitHub required status checks in the  branch protection settings. But this is annoying, because you can't tel GitHub to require the overall workflow to pass. You can only tell GitHub the names of each of the individual jobs. So whenever a job is created, removed, renamed, or has some of the matrix variables changed, you had to go into the repo settings and update the branch protection settings. This is supremely annoying.

So the idea here is to have a job named `finalize`. We mark this job as the only required status check. And then, when we add or edit jobs, we just update the `needs` dependencies of the `finalize` job.

However, we have to include the `if: always()` clause, because if we don't, then the `finalize` job will be marked as "skipped". But for GitHub branch protection settings, if a job is marked as "skipped", GitHub considers that to be a pass!!![^1] So then GitHub allows you to merge the PR, even though `finalize` is marked as "skipped".

[^1]: This is supremely annoying. I think there is a GitHub discussion about it somewhere, but IIRC a GitHub employee said that this is the intended behavior, and they won't be changing it.